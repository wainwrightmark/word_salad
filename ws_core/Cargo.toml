[package]
name = "ws_core"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rand = "0.8.5"
itertools = "0.12"
strum = { version = "0.25", features = ["derive"] }
chrono = { version = "0.4", features = ["serde", "wasmbind"] }
serde = { version = "1", features = ["derive", "rc"] }
anyhow = "1"
base64 = "0.21"
geometrid = { git = "https://github.com/wainwrightmark/geometrid", features = ["glam", "serde"] }
glam = "0.24"
prime_bag = { version = "0.2.0"}
serde_with = "3"
lazy_static = "1.4"
serde_yaml = "0.9"
serde_repr = "0.1"
enumset = { version = "1", features = ["serde"] }
arrayvec = { version = "0.7.4", features = ["serde"] }
num-traits = "0.2"
num-derive = "0.4"
unicode-normalization = "0.1"
unicode-segmentation = "1.10"
log = "0.4.20"
ustr = {version = "1.0", features = ["serde"]}
const_sized_bit_set ={git="https://github.com/wainwrightmark/const_sized_bit_set"}
static_assertions = "1.1.0"


[dev-dependencies]
insta = "1"
test-case = "3"
criterion = "0.5"
iai-callgrind = "0.10.0"


[[bench]]
name = "criterion_benchmark"
harness = false

[[bench]]
name = "iai_benchmark"
harness = false