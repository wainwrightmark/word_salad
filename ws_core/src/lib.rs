pub mod character;
pub mod colors;
pub mod designed_level;
pub mod display_word;
pub mod finder;
pub mod layout;
pub mod word;
pub mod level_type;
pub use crate::prelude::*;

pub mod prelude {

    pub use crate::character::*;
    pub use crate::colors::*;
    pub use crate::designed_level::*;
    pub use crate::display_word::*;
    pub use crate::word::*;

    pub use arrayvec::ArrayVec;
    pub use geometrid::prelude::HasCenter;
    pub use geometrid::prelude::*;
    pub use std::array;
    pub use ustr::Ustr;

    pub use crate::layout::prelude::*;

    pub type Tile = geometrid::tile::Tile<4, 4>;

    pub type CharsArray = ArrayVec<Character, 16>;
    pub type Grid = geometrid::tile_map::TileMap<Character, 4, 4, 16>;
    pub type GridSet = geometrid::tile_set::TileSet16<4, 4, 16>;
    // pub type Vertex = geometrid::vertex::Vertex<4, 4>;
    pub type Solution = ArrayVec<Tile, 16>;

    pub fn try_make_grid(text: &str) -> Option<Grid> {
        let mut arr = [Character::Blank; 16];
        for (index, char) in text.chars().enumerate() {
            let c = Character::try_from(char).ok()?;
            *arr.get_mut(index)? = c;
        }

        Some(Grid::from_inner(arr))
    }
}
