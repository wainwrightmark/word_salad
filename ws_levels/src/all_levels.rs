use itertools::Itertools;
use lazy_static::lazy_static;
use ws_core::{DesignedLevel, Numbering, Ustr};

lazy_static! { //todo data_bake
    pub(crate) static ref TUTORIAL: Vec<DesignedLevel> = include_str!("levels/tutorial.tsv")
        .lines()
        .map(DesignedLevel::from_tsv_line)
        .map(|x| x.unwrap())
        .collect_vec();


        pub(crate) static ref US_STATES: Vec<DesignedLevel> = number_levels(
            include_str!("levels/geography/us_states.tsv")
                .lines()
                .map(DesignedLevel::from_tsv_line)
                .map(|x| x.unwrap()),
            "US States"
        );

    pub(crate) static ref EUROPEAN_CAPITALS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/geography/european_capitals.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "European Capitals"
    );
    pub(crate) static ref EUROPEAN_COUNTRIES: Vec<DesignedLevel> = number_levels(
        include_str!("levels/geography/european_countries.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "European Countries"
    );


    pub(crate) static ref SOUTH_AND_EAST_ASIAN_COUNTRIES: Vec<DesignedLevel> = number_levels(
        include_str!("levels/geography/south_and_east_asian_countries.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "South & East Asian Countries"
    );

    pub(crate) static ref MIDDLE_EASTERN_COUNTRIES: Vec<DesignedLevel> = number_levels(
        include_str!("levels/geography/middle_eastern_countries.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "Middle Eastern Countries"
    );

    pub(crate) static ref AFRICAN_COUNTRIES: Vec<DesignedLevel> = number_levels(
        include_str!("levels/geography/african_countries.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "African Countries"
    );

    pub(crate) static ref SOUTH_AND_EAST_ASIAN_CAPITALS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/geography/south_and_east_asian_capitals.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "South & East Asian Capitals"
    );

    pub(crate) static ref MIDDLE_EASTERN_CAPITALS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/geography/middle_eastern_capitals.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "Middle Eastern Capitals"
    );




    pub(crate) static ref INSECTS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/natural_world/insects.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "Insects"
    );
    pub(crate) static ref FRUIT: Vec<DesignedLevel> = number_levels(
        include_str!("levels/natural_world/fruit.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "Fruit"
    );
    pub(crate) static ref GEMSTONES: Vec<DesignedLevel> = number_levels(
        include_str!("levels/natural_world/gemstones.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "Gemstones"
    );
    pub(crate) static ref VEGETABLES: Vec<DesignedLevel> = number_levels(
        include_str!("levels/natural_world/vegetables.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "Vegetables"
    );
    pub(crate) static ref ELEMENTS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/natural_world/elements.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "Elements"
    );

    pub(crate) static ref MAMMALS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/natural_world/mammals.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "Mammals"
    );

    pub(crate) static ref BIRDS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/natural_world/birds.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "Birds"
    );

    pub(crate) static ref REPTILES_AND_AMPHIBIANS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/natural_world/reptiles and amphibians.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "Reptiles & Amphibians"
    );

    pub(crate) static ref NFL_TEAMS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/US Sports/NFL Teams.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "NFL Teams"
    );

    pub(crate) static ref NHL_TEAMS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/US Sports/NHL Teams.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "NHL Teams"
    );

    pub(crate) static ref NBA_TEAMS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/US Sports/NBA Teams.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "NBA Teams"
    );

    pub(crate) static ref MLB_TEAMS: Vec<DesignedLevel> = number_levels(
        include_str!("levels/US Sports/MLB Teams.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap()),
        "MLB Teams"
    );

    pub static ref DEFAULT_DAILY_CHALLENGE: Vec<DesignedLevel> = include_str!("../../daily.tsv")
            .lines()
            .map(DesignedLevel::from_tsv_line)
            .map(|x| x.unwrap())
            .collect_vec();


    pub static ref DAILY_CHALLENGE_NUMBERED: Vec<DesignedLevel> = {
        let mut levels = (*DEFAULT_DAILY_CHALLENGE).clone();
        number_daily_challenge_levels(&mut levels);
        levels

    };
}

pub fn number_daily_challenge_levels(levels: &mut Vec<DesignedLevel>) {
    for (index, level) in levels.iter_mut().enumerate() {
        level.numbering = Some(Numbering::WordSaladNumber(index + 1));
    }
}

pub fn number_levels(
    levels: impl Iterator<Item = DesignedLevel>,
    category: &'static str,
) -> Vec<DesignedLevel> {
    let mut r = vec![];
    let mut index = 1;
    for mut l in levels {
        l.numbering = Some(Numbering::SequenceNumber(index));
        l.name = Ustr::from(category);
        index += 1;
        r.push(l)
    }

    r
}

pub fn get_tutorial_level(index: usize) -> Option<&'static DesignedLevel> {
    TUTORIAL.get(index)
}

#[cfg(test)]
pub mod tests {

    use std::{collections::HashSet, str::FromStr};

    use strum::IntoEnumIterator;
    use ws_core::{
        finder::{
            cluster::*, falling_probability, helpers::FinderSingleWord, node::GridResult,
            orientation,
        },
        prelude::*,
    };

    use crate::prelude::LevelSequence;

    use super::*;

    pub fn get_all_levels() -> Vec<DesignedLevel> {
        [
            TUTORIAL.iter(),
            US_STATES.iter(),
            EUROPEAN_COUNTRIES.iter(),
            EUROPEAN_CAPITALS.iter(),
            SOUTH_AND_EAST_ASIAN_COUNTRIES.iter(),
            MIDDLE_EASTERN_COUNTRIES.iter(),
            SOUTH_AND_EAST_ASIAN_CAPITALS.iter(),
            MIDDLE_EASTERN_CAPITALS.iter(),
            INSECTS.iter(),
            FRUIT.iter(),
            GEMSTONES.iter(),
            VEGETABLES.iter(),
            ELEMENTS.iter(),
            MAMMALS.iter(),
            BIRDS.iter(),
        ]
        .iter()
        .cloned()
        .flat_map(|x| x)
        .cloned()
        .collect_vec()
    }

    #[test]
    pub fn test_all_levels_unique() {
        let mut set: HashSet<_> = Default::default();

        let mut all_errors: Vec<String> = Default::default();

        for level in get_all_levels()
            .iter()
            .chain((*DEFAULT_DAILY_CHALLENGE).iter())
        {
            let words = level
                .words
                .iter()
                .map(|x: &DisplayWord| x.characters.clone())
                .sorted()
                .collect_vec();

            if !set.insert(words) {
                all_errors.push(format!("Level '{}' is a duplicate", level.to_string()));
            }
        }

        for error in all_errors.iter() {
            println!("{error}")
        }

        assert!(all_errors.is_empty())
    }

    #[test]
    pub fn test_all_levels_valid() {
        let levels = get_all_levels();

        assert!(levels.len() > 5);

        let mut all_errors: Vec<String> = Default::default();

        for level in levels {
            let name = &level.name;
            assert!(level.words.len() > 0, "Level {name} should have words");
            for word in level.words.iter() {
                let solution = word.find_solution(&level.grid);
                if solution.is_none() {
                    panic!(
                        "Level '{name}' has no solution for '{word}'",
                        word = word.text
                    )
                }
            }

            if let Err(err) = test_grid_not_taboo(&level) {
                if !TABOO_EXCEPTIONS.contains(
                    &(level
                        .grid
                        .into_iter()
                        .map(|x| x.as_char().to_ascii_uppercase())
                        .join("")
                        .as_str()),
                ) {
                    all_errors.push(err);
                }
            }

            test_word_ordering(&level, &mut all_errors);
        }

        for error in all_errors.iter() {
            println!("{error}")
        }

        assert!(all_errors.is_empty())
    }

    #[test]
    pub fn test_daily_challenge_levels_valid() {
        assert!(DEFAULT_DAILY_CHALLENGE.len() > 10);

        let mut all_errors: Vec<String> = Default::default();

        for level in DEFAULT_DAILY_CHALLENGE.clone().into_iter() {
            let name = &level.name;
            if level.words.len() < 4 {
                all_errors.push(format!("Level {name} should have at least 4 words"))
            }

            for word in level.words.iter() {
                let solution = word.find_solution(&level.grid);
                if solution.is_none() {
                    all_errors.push(format!(
                        "Level '{name}' has no solution for '{word}'",
                        word = word.text
                    ));
                }
            }

            if let Err(err) = test_grid_not_taboo(&level) {
                if !TABOO_EXCEPTIONS.contains(
                    &(level
                        .grid
                        .into_iter()
                        .map(|x| x.as_char().to_ascii_uppercase())
                        .join("")
                        .as_str()),
                ) {
                    all_errors.push(err);
                }
            }

            test_word_ordering(&level, &mut all_errors);

            if let Some(colors) = level.special_colors {
                if colors.len() < 4 {
                    all_errors.push(format!(
                        "Level {name} has custom colors but fewer than four"
                    ));
                }
            }
        }

        for error in all_errors.iter() {
            println!("{error}")
        }

        assert!(all_errors.is_empty())
    }

    #[test]
    pub fn test_sequence_clustering() {
        let mut text: String = String::default();

        for sequence in LevelSequence::iter() {
            let cluster = Cluster::from_levels(&sequence.levels());

            text.push_str(format!("{:50} {}\n", sequence.name(), cluster.header()).as_str());
        }

        insta::assert_snapshot!(text);
    }

    #[test]
    pub fn test_daily_challenge_data() {
        let mut text = "Name\tInfo\tWord Count\tUtilization\tChance to fall after 1st word\tChance to fall after 2nd word\tChance to have fallen after 2 words\tWords\n".to_string();

        for level in (*DEFAULT_DAILY_CHALLENGE).iter() {
            let fa1 = falling_probability::calculate_falling_probability_1(level) * 100.0;
            let fa2 = falling_probability::calculate_falling_probability_2(level) * 100.0;
            let fac2 =
                falling_probability::calculate_cumulative_falling_probability_2(level) * 100.0;
            let words_count = level.words.len();
            let utilization = level
                .words
                .iter()
                .map(|x| x.characters.len())
                .sum::<usize>() as f32
                / level.grid.iter().filter(|x| !x.is_blank()).count() as f32;
            let data =
                format!("{words_count:2}\t{utilization:1.2}\t{fa1:3.1}%\t{fa2:3.1}%\t{fac2:3.1}%");

            let words = level.words.iter().map(|x| x.text).join("\t");

            text.push_str(
                format!(
                    "{name:30}\t{info:20}\t{data}\t{words}\n",
                    name = level.full_name().as_str(),
                    info = level.extra_info.unwrap_or_default().as_str()
                )
                .as_str(),
            );
        }

        insta::assert_snapshot!(text);
    }

    fn test_word_ordering(level: &DesignedLevel, errors: &mut Vec<String>) {
        for (a, b) in level.words.iter().tuple_windows() {
            if a > b {
                errors.push(format!("{b} should come before {a}"));
            }
        }
    }

    fn test_grid_not_taboo(level: &DesignedLevel) -> Result<(), String> {
        if let Some(taboo_word) = orientation::find_taboo_word(&level.grid) {
            let mut gr = GridResult {
                grid: level.grid,
                words: level
                    .words
                    .iter()
                    .map(|x| FinderSingleWord::from_str(&x.text).unwrap())
                    .collect_vec(),
                letters: Default::default(), //doesn't matter
            };

            let optimize_result = orientation::try_optimize_orientation(&mut gr);

            match optimize_result {
                Ok(_) => {
                    return Err(format!(
                        "Level '{:<26}' Grid '{:?}' contains taboo word '{taboo_word:?}'. Try {}",
                        level.name.to_string(),
                        level.grid.iter().join(""),
                        gr.grid.iter().join("")
                    ));
                }
                Err(message) => {
                    return Err(format!(
                        "Level '{:<26}' Grid '{:?}' contains taboo word '{taboo_word:?}'. {}",
                        level.name.to_string(),
                        level.grid.iter().join(""),
                        message
                    ));
                }
            }
        }
        Ok(())
    }

    //spellchecker:disable
    pub const TABOO_EXCEPTIONS: &[&str] = &["ZEUTMSEIORHDAPON"];
    //spellchecker:enable
}
