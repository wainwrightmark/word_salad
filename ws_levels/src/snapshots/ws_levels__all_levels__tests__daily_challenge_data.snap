---
source: ws_levels/src/all_levels.rs
expression: text
---
Name	Info	Word Count	Utilization	Chance to fall after 1st word	Chance to fall after 2nd word	Chance to have fallen after 2 words	Words
Packing for School            	by samuel           	 6	2.06	66.7%	73.3%	100.0%	Apple	Eraser	Folder	Paper	Pencil	Ruler
Elements                      	by mark             	11	3.81	27.3%	31.8%	52.7%	Argon	Arsenic	Barium	Boron	Bromine	Carbon	Iodine	Iron	Neon	Radium	Radon
US States                     	by mark             	 7	2.50	28.6%	57.1%	71.4%	Florida	Idaho	Iowa	Maine	Maryland	New York	Ohio
Signs of the Zodiac           	by samuel           	 5	2.19	80.0%	90.0%	100.0%	Aquarius	Aries	Sagittarius	Taurus	Virgo
Marvel's Avengers             	by mark             	 5	1.69	60.0%	90.0%	100.0%	Ant-Man	Hulk	Iron Man	Thor	Vision
South American Capitals       	by samuel           	 5	1.94	100.0%	100.0%	100.0%	Bogota	Brasilia	Lima	Quito	Santiago
Great American Novelists      	by samuel           	 5	1.88	40.0%	80.0%	90.0%	Bellow	Heller	Hemingway	Roth	Twain
Famous Queens                 	by samuel           	 5	1.88	80.0%	85.0%	100.0%	Anne	Caroline	Elizabeth	Helen	Mary
Classic Game Characters       	by mark             	 5	1.93	80.0%	90.0%	100.0%	Lara Croft	Link	Mario	Pac Man	Sonic
Ivy League Schools            	by mark             	 5	1.75	100.0%	100.0%	100.0%	Brown	Columbia	Cornell	Penn	Yale
Insects                       	by mark             	 7	2.44	57.1%	66.7%	85.7%	Flea	Hornet	Locust	Louse	Mantis	Mosquito	Moth
Simpsons Characters           	by samuel           	 6	1.94	83.3%	86.7%	100.0%	Bart	Homer	Lisa	Maggie	Marge	Mr Burns
US States                     	by mark             	 7	2.56	28.6%	47.6%	66.7%	California	Florida	Hawaii	Idaho	Iowa	Maine	Ohio
Colors                        	by samuel           	 7	1.94	71.4%	81.0%	100.0%	Blue	Coral	Lime	Pink	Plum	Silver	Teal
Disney Movies                 	by mark             	 5	1.81	80.0%	85.0%	100.0%	Aladdin	Bambi	Dumbo	Encanto	Moana
Aquatic Mammals               	by samuel           	 6	2.06	83.3%	86.7%	100.0%	Orca	Otter	Polar Bear	Seal	Walrus	Whale
Counties of Ireland           	for St. Patrick's Day	 5	1.81	80.0%	85.0%	100.0%	Derry	Donegal	Dublin	Galway	Kerry
South American Countries      	by samuel           	 5	1.88	100.0%	100.0%	100.0%	Bolivia	Brazil	Chile	Colombia	Peru
Safari Animals                	by mark             	 6	2.44	66.7%	70.0%	93.3%	Antelope	Cheetah	Elephant	Leopard	Lion	Zebra
Poets                         	for World Poetry Day	 5	1.94	60.0%	75.0%	90.0%	Eliot	Frost	Keats	Plath	Shakespeare
Premier League Teams          	by mark             	 5	1.93	60.0%	85.0%	100.0%	Arsenal	Chelsea	Fulham	Man U	Spurs
Parts of the Body             	by samuel           	10	2.94	30.0%	33.3%	53.3%	Ankle	Cheek	Chest	Chin	Finger	Heel	Knee	Nose	Shin	Tongue
Shakespeare Characters        	for Shakespeare's Birthday	 6	2.06	50.0%	66.7%	93.3%	Iago	Juliet	Macbeth	Oberon	Romeo	Timon
Vegetables                    	by samuel           	 7	2.44	71.4%	73.8%	95.2%	Artichoke	Carrot	Kale	Leek	Okra	Potato	Tomato
Asian Countries               	by mark             	 7	2.56	42.9%	59.5%	85.7%	China	India	Indonesia	Japan	Laos	Nepal	Thailand
Animals that might eat you    	by samuel           	 6	1.81	83.3%	90.0%	100.0%	Bear	Hyena	Lion	Piranha	Shark	Wolf
Metals                        	by mark             	 7	2.12	85.7%	88.1%	100.0%	Brass	Bronze	Cobalt	Gold	Iron	Lead	Steel
F·R·I·E·N·D·S Characters      	by mark             	 5	1.75	80.0%	85.0%	100.0%	Chandler	Joey	Phoebe	Rachel	Ross
Jesus' Disciples              	for Good Friday     	 5	1.75	100.0%	95.0%	100.0%	Andrew	James	Judas	Matthew	Peter
Citrus Fruit                  	by mark             	 6	2.38	50.0%	56.7%	80.0%	Lemon	Lime	Mandarin	Orange	Pomelo	Tangerine
Cooked Eggs                   	for Easter          	 5	2.00	100.0%	100.0%	100.0%	Boiled	Deviled	Fried	Pickled	Poached
????? ?????                   	by mark             	 5	1.56	100.0%	100.0%	100.0%	Baby	Ginger	Posh	Scary	Sporty
Gemstones                     	by mark             	 8	2.44	50.0%	53.6%	78.6%	Amber	Beryl	Emerald	Onyx	Opal	Pearl	Ruby	Topaz
Car Manufacturers             	by samuel           	 6	2.50	33.3%	63.3%	80.0%	Chevrolet	Chrysler	Ford	Honda	Land Rover	Tesla
Currencies                    	by samuel           	 7	2.06	85.7%	88.1%	100.0%	Dollar	Franc	Krona	Peso	Real	Rupee	Yuan
Things that are red           	by mark             	 6	2.06	66.7%	80.0%	100.0%	Blood	Cherry	Lobster	Mars	Rooster	Ruby
Greek Gods                    	by samuel           	 6	2.19	66.7%	66.7%	93.3%	Aphrodite	Ares	Hera	Hermes	Poseidon	Zeus
Famous Fictional Couples      	by mark and samuel  	 6	1.94	66.7%	80.0%	100.0%	Homer	Juliet	Marge	Rachel	Romeo	Ross
Famous Astronomers            	for eclipse day     	 5	1.88	80.0%	90.0%	100.0%	Brahe	Drake	Gallileo	Halley	Hubble
Supernatural Creatures        	by samuel           	 7	2.44	71.4%	76.2%	95.2%	Dragon	Fairy	Golem	Hydra	Mermaid	Ogre	Vampire
Famous Scientists             	by mark             	 6	2.25	33.3%	66.7%	86.7%	Curie	Darwin	Edison	Einstein	Newton	Tesla
Harry Potter Characters       	by samuel           	 5	1.75	100.0%	100.0%	100.0%	Fred	Hagrid	Harry	Hermione	Snape
African Countries             	by mark             	 5	1.75	80.0%	90.0%	100.0%	Chad	Egypt	Ethiopia	Ghana	Rwanda
Pizza Toppings                	by samuel           	 6	2.62	50.0%	60.0%	86.7%	Jalapeno	Olive	Onion	Pepper	Pepperoni	Pineapple
Time Periods                  	by samuel           	 6	1.94	83.3%	90.0%	100.0%	Decade	Hour	Moment	Month	Second	Year
Breakfast Foods               	by samuel           	 6	2.06	50.0%	76.7%	93.3%	Bacon	Bagel	Cereal	Eggs	Granola	Waffle
Birds of Prey                 	by mark             	 5	1.69	80.0%	95.0%	100.0%	Barn Owl	Eagle	Harrier	Hawk	Kite
2D Shapes                     	by mark             	 7	2.62	71.4%	73.8%	95.2%	Circle	Kite	Octagon	Oval	Rectangle	Star	Triangle
US Presidents                 	by samuel           	 8	2.56	37.5%	53.6%	75.0%	Adams	Bush	Ford	Grant	Harding	Obama	Truman	Trump
Music Genres                  	by mark             	 6	1.69	83.3%	90.0%	100.0%	Blues	Funk	Indie	Rock	Soul	Swing
Signs of the Zodiac           	by samuel           	 5	1.94	100.0%	100.0%	100.0%	Aquarius	Aries	Libra	Scorpio	Taurus
